migration/blocktransactions/blocktransactions.go

```diff
@@ -0,0 +1,291 @@
+		has, err := core.BlockTransactionsBucket.Has(database, blockNumber)
+		if err != nil {
+			return blockTransactions{}, false, err
+		}
+		if has {
+			return blockTransactions{}, false, nil
+		}
+	}
+
+	return blockTransactions{
+		Transactions: transactions,
+		Receipts:     receipts,
+	}, true, nil
+}
+
+func writeBlock(
+	database db.KeyValueWriter,
+	blockNumber uint64,
+	blockTransactions blockTransactions,
+) error {
+	bytes, err := encoder.Marshal(blockTransactions)
+	if err != nil {
+		return err
+	}
+
+	return rawBlockTransactions.Put(database, blockNumber, &bytes)
+}
+
+func deleteOldData(database db.KeyValueRangeDeleter, startBlock, endBlock uint64) error {
+	err := core.TransactionsByBlockNumberAndIndexBucket.Prefix().DeleteRange(
+		database,
+		startBlock,
+		endBlock+1,
+	)
+	if err != nil {
+		return err
+	}
+
+	err = core.ReceiptsByBlockNumberAndIndexBucket.Prefix().DeleteRange(
+		database,
+		startBlock,
+		endBlock+1,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createBatch(
+	logger utils.SimpleLogger,
+	database db.KeyValueStore,
+	startBlock,
+	endBlock uint64,
+) (db.Batch, error) {
+	debugLog(logger, "fetching block", startBlock, endBlock)
+	batch := database.NewBatch()
+
+	for blockNumber := startBlock; blockNumber <= endBlock; blockNumber++ {
+		blockTransactions, shouldMigrate, err := readBlock(database, blockNumber)
+		if err != nil {
+			return nil, err
+		}
+		if !shouldMigrate {
+			logger.Infow("skipping block", "blockNumber", blockNumber)
+			continue
+		}
+		if err := writeBlock(batch, blockNumber, blockTransactions); err != nil {
+			return nil, err
+		}
+	}
+
+	debugLog(logger, "fetched block", startBlock, endBlock)
+
+	err := deleteOldData(batch, startBlock, endBlock)
+	return batch, err
+}
+
+func debugLog(logger utils.SimpleLogger, msg string, startBlock, endBlock uint64) {
+	logger.Debugw(
+		msg,
+		"startBlock", startBlock,
+		"endBlock", endBlock,
+		"numGoroutines", runtime.NumGoroutine(),
+	)
+}
+
+func getFirstBlockInBucket[A any](items iter.Seq2[prefix.Entry[A], error]) (uint64, bool, error) {
+	for item, err := range items {
+		if err != nil {
+			return 0, false, err
+		}
+
+		var blockNumIndex db.BlockNumIndexKey
+		if err := blockNumIndex.UnmarshalBinary(item.Key[1:]); err != nil {
+			return 0, false, err
+		}
+		return blockNumIndex.Number, true, nil
+	}
+	return 0, false, nil
+}
+
+func getFirstBlock(database db.KeyValueReader) (uint64, bool, error) {
+	transactions, hasTransactions, err := getFirstBlockInBucket(
+		core.TransactionsByBlockNumberAndIndexBucket.Prefix().Scan(database),
+	)
+	if err != nil {
+		return 0, false, err
+	}
+
+	receipts, hasReceipts, err := getFirstBlockInBucket(
+		core.ReceiptsByBlockNumberAndIndexBucket.Prefix().Scan(database),
+	)
+	if err != nil {
+		return 0, false, err
+	}
+
+	if !hasReceipts && !hasTransactions {
+		return 0, false, nil
+	}
+
+	minBlock := min(transactions, receipts)
+	minBlock -= minBlock % batchSize
```

Comment:
nitpick: parentization would help a bit here in my opinion

Reply:
It's mainly to "normalize" the batch start and end to be multiples of `batchSize`, which is currently `10`. The main benefits are to guarantee that multiple runs use consistent batches. For example, we always run batch 0 -> 9, 10 -> 19, 20 -> 29,... but never get batch 5 -> 14. This is to avoid unexpected batch that isn't created during the experiments, which may  create unexpected spike in resource usage.